In [132]:
import pandas as pd
import numpy as np

In [133]:
dataset=pd.read_csv("Car_Rating data.csv")

In [134]:
dataset

Index                           Car                      Price  \
0        0        Maruti Suzuki Alto K10           3.99 - 5.96 Lakh   
1        1        Maruti Suzuki Alto 800           3.25 - 5.12 Lakh   
2        2                  Renault Kwid           4.70 - 6.45 Lakh   
3        3        Maruti Suzuki Alto K10           3.99 - 5.96 Lakh   
4        4        Maruti Suzuki Alto 800           3.25 - 5.12 Lakh   
..     ...                           ...                        ...   
167    167                        BMW i7                 1.95 Crore   
168    168  Land Rover Range Rover Sport          1.64 - 1.84 Crore   
169    169              Aston Martin DBX          3.82 - 4.63 Crore   
170    170                 Porsche Macan    88.06 Lakh - 1.53 Crore   
171    171                        BMW XM                 2.60 Crore   

     Overall Rating  Exterior  Interior  Ride Quality   Color  Airbags  \
0               4.4       4.4       4.4           4.4  Silver       12   
1               4.2       4.0       3.8           4.2   Black        8   
2               3.5       2.7       3.0           2.3   Black        2   
3               4.4       4.4       4.4           4.4   White        0   
4               4.2       4.0       3.8           4.2  Silver        4   
..              ...       ...       ...           ...     ...      ...   
167             0.0       NaN       NaN           NaN   White        4   
168             0.0       NaN       NaN           NaN  Silver        8   
169             0.0       NaN       NaN           NaN    Blue        4   
170             0.0       NaN       NaN           NaN   Black       12   
171             0.0       NaN       NaN           NaN    Blue        0   

    Fuel type  Price.1  
0      Hybrid    13328  
1      Petrol    16621  
2      Petrol     8467  
3      Hybrid     3607  
4      Petrol    11726  
..        ...      ...  
167    Petrol    20005  
168    Petrol    36110  
169    Petrol     7840  
170    Petrol      470  
171    Petrol     1300  

[172 rows x 11 columns]

In [135]:
dataset.isnull().sum()

Index              0
Car                0
Price              0
Overall Rating     0
Exterior          59
Interior          59
Ride Quality      59
Color              0
Airbags            0
Fuel type          0
Price.1            0
dtype: int64

In [ ]:
# Assuming 'dataset' is your DataFrame and you want to fill missing values in the column  with 0
dataset["Exterior"].fillna(0, inplace=True)
dataset["Interior"].fillna(0, inplace=True)
dataset["Ride Quality"].fillna(0, inplace=True)

In [95]:
dataset.isnull().sum()

Index             0
Car               0
Price             0
Overall Rating    0
Exterior          0
Interior          0
Ride Quality      0
Color             0
Airbags           0
Fuel type         0
Price.1           0
dtype: int64

In [96]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 172 entries, 0 to 171
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Index           172 non-null    int64  
 1   Car             172 non-null    object 
 2   Price           172 non-null    object 
 3   Overall Rating  172 non-null    float64
 4   Exterior        172 non-null    float64
 5   Interior        172 non-null    float64
 6   Ride Quality    172 non-null    float64
 7   Color           172 non-null    object 
 8   Airbags         172 non-null    int64  
 9   Fuel type       172 non-null    object 
 10  Price.1         172 non-null    int64  
dtypes: float64(4), int64(3), object(4)
memory usage: 14.9+ KB


In [97]:
dataset.columns

Index(['Index', 'Car', 'Price', 'Overall Rating', 'Exterior', 'Interior',
       'Ride Quality', 'Color', 'Airbags', 'Fuel type', 'Price.1'],
      dtype='object')

In [98]:
dir(dataset)

['Airbags',
 'Car',
 'Color',
 'Exterior',
 'Index',
 'Interior',
 'Price',
 'T',
 '_AXIS_LEN',
 '_AXIS_ORDERS',
 '_AXIS_TO_AXIS_NUMBER',
 '_HANDLED_TYPES',
 '__abs__',
 '__add__',
 '__and__',
 '__annotations__',
 '__array__',
 '__array_priority__',
 '__array_ufunc__',
 '__bool__',
 '__class__',
 '__contains__',
 '__copy__',
 '__dataframe__',
 '__dataframe_consortium_standard__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__or__',


In [99]:
dataset["Ride Quality"].dtypes

dtype('float64')

In [100]:
dataset["Car"].dtypes

dtype('O')

In [101]:
for columnname in dataset.columns:
    print(columnname)

Index
Car
Price
Overall Rating
Exterior
Interior
Ride Quality
Color
Airbags
Fuel type
Price.1


In [102]:
if dataset["Ride Quality"].dtype=="O":
    print("qual")
else:
    print("quan")

quan


In [103]:
quan=[]
qual=[]
for columnname in dataset.columns:
    if (dataset[columnname].dtype=="O"):
        #print("qual")
         qual.append(columnname)
    else:
        #print("quan")
         quan.append(columnname)

In [104]:
quan

['Index',
 'Overall Rating',
 'Exterior',
 'Interior',
 'Ride Quality',
 'Airbags',
 'Price.1']

In [105]:
qual

['Car', 'Price', 'Color', 'Fuel type']

In [106]:
def quanQual(dataset):
    quan=[]
    qual=[]
    for columnName in dataset.columns:
        #print(columnName)
        if dataset[columnName].dtype=="O":
            #print("quan")
            quan.append(columnName)
        else:
            #print("qual")
            qual.append(columnName)
    return quan,qual

In [107]:
quanQual(dataset)

(['Car', 'Price', 'Color', 'Fuel type'],
 ['Index',
  'Overall Rating',
  'Exterior',
  'Interior',
  'Ride Quality',
  'Airbags',
  'Price.1'])

In [108]:
qual,quan=quanQual(dataset)

In [109]:
qual

['Car', 'Price', 'Color', 'Fuel type']

In [110]:
quan

['Index',
 'Overall Rating',
 'Exterior',
 'Interior',
 'Ride Quality',
 'Airbags',
 'Price.1']

In [111]:
dataset

Index                           Car                      Price  \
0        0        Maruti Suzuki Alto K10           3.99 - 5.96 Lakh   
1        1        Maruti Suzuki Alto 800           3.25 - 5.12 Lakh   
2        2                  Renault Kwid           4.70 - 6.45 Lakh   
3        3        Maruti Suzuki Alto K10           3.99 - 5.96 Lakh   
4        4        Maruti Suzuki Alto 800           3.25 - 5.12 Lakh   
..     ...                           ...                        ...   
167    167                        BMW i7                 1.95 Crore   
168    168  Land Rover Range Rover Sport          1.64 - 1.84 Crore   
169    169              Aston Martin DBX          3.82 - 4.63 Crore   
170    170                 Porsche Macan    88.06 Lakh - 1.53 Crore   
171    171                        BMW XM                 2.60 Crore   

     Overall Rating  Exterior  Interior  Ride Quality   Color  Airbags  \
0               4.4       4.4       4.4           4.4  Silver       12   
1               4.2       4.0       3.8           4.2   Black        8   
2               3.5       2.7       3.0           2.3   Black        2   
3               4.4       4.4       4.4           4.4   White        0   
4               4.2       4.0       3.8           4.2  Silver        4   
..              ...       ...       ...           ...     ...      ...   
167             0.0       0.0       0.0           0.0   White        4   
168             0.0       0.0       0.0           0.0  Silver        8   
169             0.0       0.0       0.0           0.0    Blue        4   
170             0.0       0.0       0.0           0.0   Black       12   
171             0.0       0.0       0.0           0.0    Blue        0   

    Fuel type  Price.1  
0      Hybrid    13328  
1      Petrol    16621  
2      Petrol     8467  
3      Hybrid     3607  
4      Petrol    11726  
..        ...      ...  
167    Petrol    20005  
168    Petrol    36110  
169    Petrol     7840  
170    Petrol      470  
171    Petrol     1300  

[172 rows x 11 columns]

# CENTRAL TENDENCY
# REMOVING OF OUTLIER LOW AND HIGHER (NONE,NULL)
# MEAN:AVERAGE WHIVH GIVES [OVERALL CENTER VALUE]----2.89%
# MEDIAN:MIDPOINT OF DATA[(N/2),(N+1/2)] (N-VALUES ISEVEN)----4.2%
# MODE REPEATED DATAPOINT-----0%

In [112]:
dataset["Ride Quality"].mean()

2.8930232558139535

In [113]:
dataset["Ride Quality"].median()

4.2

In [114]:
dataset["Ride Quality"].mode()[0]

0.0

In [115]:
dataset.describe()

Index  Overall Rating    Exterior    Interior  Ride Quality  \
count  172.000000      172.000000  172.000000  172.000000    172.000000   
mean    85.500000        2.956977    2.926744    2.888372      2.893023   
std     49.796252        2.074690    2.158049    2.128471      2.143195   
min      0.000000        0.000000    0.000000    0.000000      0.000000   
25%     42.750000        0.000000    0.000000    0.000000      0.000000   
50%     85.500000        4.000000    4.200000    4.100000      4.200000   
75%    128.250000        4.500000    4.700000    4.500000      4.500000   
max    171.000000        5.000000    5.000000    5.000000      5.000000   

          Airbags       Price.1  
count  172.000000    172.000000  
mean     6.622093  16164.593023  
std      4.454078  16608.107786  
min      0.000000     30.000000  
25%      4.000000   3102.000000  
50%      6.000000  13015.000000  
75%     12.000000  21051.250000  
max     16.000000  87112.000000

In [116]:
descriptive=pd.DataFrame()

In [117]:
descriptive

Empty DataFrame
Columns: []
Index: []

In [118]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode"],columns=quan)

In [119]:
descriptive

Index Overall Rating Exterior Interior Ride Quality Airbags Price.1
Mean     NaN            NaN      NaN      NaN          NaN     NaN     NaN
Median   NaN            NaN      NaN      NaN          NaN     NaN     NaN
Mode     NaN            NaN      NaN      NaN          NaN     NaN     NaN

In [120]:
descriptive["Ride Quality"]["median"]=0

In [121]:
descriptive

Index Overall Rating Exterior Interior Ride Quality Airbags Price.1
Mean     NaN            NaN      NaN      NaN          NaN     NaN     NaN
Median   NaN            NaN      NaN      NaN          NaN     NaN     NaN
Mode     NaN            NaN      NaN      NaN          NaN     NaN     NaN

In [126]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()

In [127]:
descriptive

Index Overall Rating  Exterior  Interior Ride Quality   Airbags  \
Mean    85.5       2.956977  2.926744  2.888372     2.893023  6.622093   
Median  85.5            4.0       4.2       4.1          4.2       6.0   
Mode     NaN            NaN       NaN       NaN          NaN       NaN   

             Price.1  
Mean    16164.593023  
Median       13015.0  
Mode             NaN

In [128]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()
    descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]

# 1 (Overall Rating_):average
# Mean: 2.956977
# Median: 4.0
# Mode: 0.0
# Average: (2.956977 + 4.0 + 0.0) / 3 = 2.985659
# 2 (Exterior):
# Mean: 4.454867
# Median: 4.5
# Mode:0.0
# Average: (4.454867 + 4.5 + 4.5) / 3 =4.485289
# 3 (Ride Quality):
# Mean: 4.40354
# Median: 4.5
# Mode: 0.0
# Average: (4.40354 + 4.5 + 4.5) / 3 = 4.46784778475289

In [129]:
descriptive

Index Overall Rating  Exterior  Interior Ride Quality   Airbags  \
Mean    85.5       2.956977  2.926744  2.888372     2.893023  6.622093   
Median  85.5            4.0       4.2       4.1          4.2       6.0   
Mode       0            0.0       0.0       0.0          0.0        12   

             Price.1  
Mean    16164.593023  
Median       13015.0  
Mode             549

In [35]:
np.percentile(dataset["Overall Rating"],25)

0.0

In [36]:
np.percentile(dataset["Overall Rating"],75)

4.5

In [37]:
np.percentile(dataset["Overall Rating"],100)

5.0

In [38]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode","Q1:25%","Q2:50%",
                               "Q3:75%","99%","Q4:100%"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()
    descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
    descriptive[columnName]["Q1:25%"]=dataset.describe()[columnName]["25%"]
    descriptive[columnName]["Q2:50%"]=dataset.describe()[columnName]["50%"]
    descriptive[columnName]["Q3:75%"]=dataset.describe()[columnName]["75%"]
    descriptive[columnName]["99%"]=np.percentile(dataset[columnName],99)
    descriptive[columnName]["Q4:100%"]=dataset.describe()[columnName]["max"]
    

# Q1 (25%): This represents the first quartile, which is the value below which 25% of the data falls. For example, in the "Index" variable, 25% of the data falls below 42.75.

# Q2 (50%): This represents the second quartile, which is equivalent to the median. It divides the data into two equal halves. For example, in the "Overall Rating" variable, the median value is 4.0.

# Q3 (75%): This represents the third quartile, which is the value below which 75% of the data falls. For example, in the "Exterior" variable, 75% of the data falls below 4.9.

# 99%: This represents the 99th percentile, which indicates the value below which 99% of the data falls. For example, in the "Index" variable, 99% of the data falls below 169.29.

# Q4 (100%): This represents the maximum value in the dataset for each variable. For example, in the "Index" variable, the maximum value is 171.0.

In [130]:
descriptive

Index Overall Rating  Exterior  Interior Ride Quality   Airbags  \
Mean    85.5       2.956977  2.926744  2.888372     2.893023  6.622093   
Median  85.5            4.0       4.2       4.1          4.2       6.0   
Mode       0            0.0       0.0       0.0          0.0        12   

             Price.1  
Mean    16164.593023  
Median       13015.0  
Mode             549

In [40]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode","Q1:25%","Q2:50%",
                               "Q3:75%","99%","Q4:100%","IQR","1.5rule","Lesser","Greater","Min","Max"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()
    descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
    descriptive[columnName]["Q1:25%"]=dataset.describe()[columnName]["25%"]
    descriptive[columnName]["Q2:50%"]=dataset.describe()[columnName]["50%"]
    descriptive[columnName]["Q3:75%"]=dataset.describe()[columnName]["75%"]
    descriptive[columnName]["99%"]=np.percentile(dataset[columnName],99)
    descriptive[columnName]["Q4:100%"]=dataset.describe()[columnName]["max"]
    descriptive[columnName]["IQR"]=descriptive[columnName]["Q3:75%"]-descriptive[columnName]["Q1:25%"]
    descriptive[columnName]["1.5rule"]=1.5*descriptive[columnName]["IQR"]
    descriptive[columnName]["Lesser"]=descriptive[columnName]["Q1:25%"]-descriptive[columnName]["1.5rule"]
    descriptive[columnName]["Greater"]=descriptive[columnName]["Q3:75%"]+descriptive[columnName]["1.5rule"]
    descriptive[columnName]["Min"]=dataset[columnName].min()
    descriptive[columnName]["Max"]=dataset[columnName].max()
    

# IQR (Interquartile Range): This represents the range between the first quartile (Q1) and the third quartile (Q3). It provides a measure of the spread of the middle 50% of the data. For example:
# Index: IQR = 85.5
# Overall Rating: IQR = 4.5
# Exterior: IQR = 0.7
# Interior: IQR = 0.7
# Ride Quality: IQR = 0.6

# 1.5 * IQR rule: This rule is used to identify potential outliers in a dataset. It states that any data point that is:
# Lesser than Q1 - 1.5 * IQR or
# Greater than Q3 + 1.5 * IQRIQR

# Lesser: This represents the lower bound for identifying potential outliers using the 1.5 * IQR rule. For example:

# Index: Lesser = Q1 - 1.5 * IQR = -85.5
# Overall Rating: Lesser = Q1 - 1.5 * IQR = -6.75
# Exterior: Lesser = Q1 - 1.5 * IQR = 3.15
# Interior: Lesser = Q1 - 1.5 * IQR = 3.05
# Ride Quality: Lesser = Q1 - 1.5 * IQR = 3.3

# Greater: This represents the upper bound for identifying potential outliers using the 1.5 * IQR rule. For example:

# Index: Greater = Q3 + 1.5 * IQR = 256.5
# Overall Rating: Greater = Q3 + 1.5 * IQR = 11.25
# Exterior: Greater = Q3 + 1.5 * IQR = 5.95
# Interior: Greater = Q3 + 1.5 * IQR = 5.85
# Ride Quality: Greater = Q3 + 1.5 * IQR = 5.7

# the IQR and the 1.5 * IQR rule are useful tools in exploratory data analysis to identify potential outliers in a dataset. They help in understanding the distribution and variability of the data. 

In [131]:
descriptive

Index Overall Rating  Exterior  Interior Ride Quality   Airbags  \
Mean    85.5       2.956977  2.926744  2.888372     2.893023  6.622093   
Median  85.5            4.0       4.2       4.1          4.2       6.0   
Mode       0            0.0       0.0       0.0          0.0        12   

             Price.1  
Mean    16164.593023  
Median       13015.0  
Mode             549

In [42]:
descriptive["Overall Rating"]["Lesser"]<descriptive["Ride Quality"]["Min"]

True

In [44]:
descriptive["Overall Rating"]["Lesser"]>descriptive["Ride Quality"]["Min"]

False

In [45]:
Lesser=[]
Greater=[]
for columnName in quan:
    if(descriptive[columnName]["Min"]<descriptive[columnName]["Lesser"]):
        Lesser.append(columnName)
    if(descriptive[columnName]["Max"]>descriptive[columnName]["Greater"]):
        Greater.append(columnName)

In [46]:
Lesser

['Exterior', 'Interior', 'Ride Quality']

In [47]:
Greater

[]

In [48]:
for columnName in Lesser:
    dataset[columnName][dataset[columnName]<descriptive[columnName]["Lesser"]]=descriptive[columnName]["Lesser"]
for columnName in Greater:
    dataset[columnName][dataset[columnName]>descriptive[columnName]["Greater"]]=descriptive[columnName]["Greater"]

C:\Users\ragus\AppData\Local\Temp\ipykernel_19028\2761044072.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset[columnName][dataset[columnName]<descriptive[columnName]["Lesser"]]=descriptive[columnName]["Lesser"]
C:\Users\ragus\AppData\Local\Temp\ipykernel_19028\2761044072.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset[columnName][dataset[columnName]<descriptive[columnName]["Lesser"]]=descriptive[columnName]["Lesser"]
C:\Users\ragus\AppData\Local\Temp\ipykernel_19028\2761044072.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the 

In [49]:
def Univariate(dataset, quan):
    descriptive=pd.DataFrame(index=["Mean","Median","Mode","Q1:25%","Q2:50%",
                                   "Q3:75%","99%","Q4:100%","IQR","1.5rule","Lesser","Greater","Min","Max"],columns=quan)
    for columnName in quan:
        descriptive[columnName]["Mean"]=dataset[columnName].mean()
        descriptive[columnName]["Median"]=dataset[columnName].median()
        descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
        descriptive[columnName]["Q1:25%"]=dataset.describe()[columnName]["25%"]
        descriptive[columnName]["Q2:50%"]=dataset.describe()[columnName]["50%"]
        descriptive[columnName]["Q3:75%"]=dataset.describe()[columnName]["75%"]
        descriptive[columnName]["99%"]=np.percentile(dataset[columnName],99)
        descriptive[columnName]["Q4:100%"]=dataset.describe()[columnName]["max"]
        descriptive[columnName]["IQR"]=descriptive[columnName]["Q3:75%"]-descriptive[columnName]["Q1:25%"]
        descriptive[columnName]["1.5rule"]=1.5*descriptive[columnName]["IQR"]
        descriptive[columnName]["Lesser"]=descriptive[columnName]["Q1:25%"]-descriptive[columnName]["1.5rule"]
        descriptive[columnName]["Greater"]=descriptive[columnName]["Q3:75%"]+descriptive[columnName]["1.5rule"]
        descriptive[columnName]["Min"]=dataset[columnName].min()
        descriptive[columnName]["Max"]=dataset[columnName].max()
    return descriptive  

In [50]:
descriptive

Index Overall Rating  Exterior Interior Ride Quality
Mean       85.5       2.956977  4.454867  4.39646      4.40354
Median     85.5            4.0       4.5      4.4          4.5
Mode          0            0.0       4.5      5.0          4.5
Q1:25%    42.75            0.0       4.2      4.1          4.2
Q2:50%     85.5            4.0       4.5      4.4          4.5
Q3:75%   128.25            4.5       4.9      4.8          4.8
99%      169.29            5.0       NaN      NaN          NaN
Q4:100%   171.0            5.0       5.0      5.0          5.0
IQR        85.5            4.5       0.7      0.7          0.6
1.5rule  128.25           6.75      1.05     1.05          0.9
Lesser    -85.5          -6.75      3.15     3.05          3.3
Greater   256.5          11.25      5.95     5.85          5.7
Min           0            0.0       2.7      3.0          2.3
Max         171            5.0       5.0      5.0          5.0

In [51]:
lesser=[]
greater=[]

for columnName in quan:
    if(descriptive[columnName]["Min"]<descriptive[columnName]["Lesser"]):
        Lesser.append(columnName)
    if(descriptive[columnName]["Max"]>descriptive[columnName]["Greater"]):
        Greater.append(columnName)        

In [52]:
lesser

[]

In [53]:
greater

[]

In [54]:
descriptive

Index Overall Rating  Exterior Interior Ride Quality
Mean       85.5       2.956977  4.454867  4.39646      4.40354
Median     85.5            4.0       4.5      4.4          4.5
Mode          0            0.0       4.5      5.0          4.5
Q1:25%    42.75            0.0       4.2      4.1          4.2
Q2:50%     85.5            4.0       4.5      4.4          4.5
Q3:75%   128.25            4.5       4.9      4.8          4.8
99%      169.29            5.0       NaN      NaN          NaN
Q4:100%   171.0            5.0       5.0      5.0          5.0
IQR        85.5            4.5       0.7      0.7          0.6
1.5rule  128.25           6.75      1.05     1.05          0.9
Lesser    -85.5          -6.75      3.15     3.05          3.3
Greater   256.5          11.25      5.95     5.85          5.7
Min           0            0.0       2.7      3.0          2.3
Max         171            5.0       5.0      5.0          5.0

# The code calculates the unique values in the "Ride Quality" column of the dataset and prints them in the order of their frequency (from most frequent to least frequent).

In [55]:
dataset["Ride Quality"].value_counts().index

Index([               4.5,                5.0,                4.2,
                      4.3,                4.7,                4.0,
                      4.4,                4.8, 3.3000000000000007,
                      4.9,                3.7,                4.1,
                      3.4,                4.6],
      dtype='float64', name='Ride Quality')

In [58]:
dataset["Overall Rating"].value_counts().values

array([55, 23, 11, 10, 10, 10,  9,  6,  6,  5,  5,  5,  4,  4,  3,  2,  2,
        1,  1], dtype=int64)

In [59]:
dataset["Overall Rating"].value_counts().sort_index

<bound method Series.sort_index of Overall Rating
0.0    55
5.0    23
4.0    11
4.1    10
4.2    10
4.3    10
4.4     9
3.9     6
4.9     6
4.5     5
4.8     5
3.5     5
4.7     4
3.7     4
4.6     3
3.2     2
3.3     2
3.0     1
3.8     1
Name: count, dtype: int64>

In [60]:
dataset["Overall Rating"].value_counts().sort_values

<bound method Series.sort_values of Overall Rating
0.0    55
5.0    23
4.0    11
4.1    10
4.2    10
4.3    10
4.4     9
3.9     6
4.9     6
4.5     5
4.8     5
3.5     5
4.7     4
3.7     4
4.6     3
3.2     2
3.3     2
3.0     1
3.8     1
Name: count, dtype: int64>

In [61]:
frequencytables=pd.DataFrame(columns=["uniqe_values","frequency","Relative Frequency","cumsum"])

In [62]:
frequencytables

Empty DataFrame
Columns: [uniqe_values, frequency, Relative Frequency, cumsum]
Index: []

In [63]:
frequencytables = pd.DataFrame(columns=["unique_values", "frequency", "Relative Frequency", "cumsum"])
frequencytables["unique_values"] = dataset["Overall Rating"].value_counts().index

In [64]:
frequencytables

unique_values frequency Relative Frequency cumsum
0             0.0       NaN                NaN    NaN
1             5.0       NaN                NaN    NaN
2             4.0       NaN                NaN    NaN
3             4.1       NaN                NaN    NaN
4             4.2       NaN                NaN    NaN
5             4.3       NaN                NaN    NaN
6             4.4       NaN                NaN    NaN
7             3.9       NaN                NaN    NaN
8             4.9       NaN                NaN    NaN
9             4.5       NaN                NaN    NaN
10            4.8       NaN                NaN    NaN
11            3.5       NaN                NaN    NaN
12            4.7       NaN                NaN    NaN
13            3.7       NaN                NaN    NaN
14            4.6       NaN                NaN    NaN
15            3.2       NaN                NaN    NaN
16            3.3       NaN                NaN    NaN
17            3.0       NaN                NaN    NaN
18            3.8       NaN                NaN    NaN

In [66]:
frequencytables["unique_values"] =dataset["Overall Rating"].value_counts().index
frequencytables["frequency"] = dataset["Overall Rating"].value_counts().values
frequencytables["relative_frequency"] = (frequencytables["frequency"] / 215)
frequencytables["cusum"] = frequencytables["relative_frequency"].cumsum()

In [67]:
def frequencytables(columnName,dataset):
    frequencytables = pd.DataFrame(columns=["unique_values", "frequency", "Relative Frequency", "cumsum"])
    frequencytables["unique_values"] =dataset[columnName].value_counts().index
    frequencytables["frequency"] = dataset[columnName].value_counts().values
    frequencytables["Relative Frequency"] = (frequencytables["frequency"] / 103)
    frequencytables["cumsum"] = frequencytables["Relative Frequency"].cumsum()
    return frequencytables

# The function frequencytables calculates frequency statistics for a given column in a dataset. It returns a DataFrame containing unique values, their frequency, relative frequency, and cumulative sum of relative frequencies.

In [68]:
frequencytables(columnName,dataset)

unique_values  frequency  Relative Frequency    cumsum
0             4.5         21            0.203883  0.203883
1             5.0         21            0.203883  0.407767
2             4.2         12            0.116505  0.524272
3             4.3          9            0.087379  0.611650
4             4.7          9            0.087379  0.699029
5             4.0          7            0.067961  0.766990
6             4.4          6            0.058252  0.825243
7             4.8          6            0.058252  0.883495
8             3.3          4            0.038835  0.922330
9             4.9          4            0.038835  0.961165
10            3.7          4            0.038835  1.000000
11            4.1          4            0.038835  1.038835
12            3.4          4            0.038835  1.077670
13            4.6          2            0.019417  1.097087

# unique_values: These are the distinct values in the "Overall Rating" column.
# frequency: This shows how often each unique value appears in the dataset.
# Relative Frequency: The proportion of each frequency to the total number of data points (103). For example, 21 out of 103 data points have a rating of 4.5, so the relative frequency is 10321≈0.203883
# cumsum: The cumulative sum of relative frequencies. It represents the proportion of data points up to and including each unique value. For example, the cumulative sum for a rating of 4.5 is 0.203883. Adding the relative frequency of 5.0 (another 0.203883) gives 0.407767,67,

In [69]:
frequencytables("Overall Rating",dataset)

unique_values  frequency  Relative Frequency    cumsum
0             0.0         55            0.533981  0.533981
1             5.0         23            0.223301  0.757282
2             4.0         11            0.106796  0.864078
3             4.1         10            0.097087  0.961165
4             4.2         10            0.097087  1.058252
5             4.3         10            0.097087  1.155340
6             4.4          9            0.087379  1.242718
7             3.9          6            0.058252  1.300971
8             4.9          6            0.058252  1.359223
9             4.5          5            0.048544  1.407767
10            4.8          5            0.048544  1.456311
11            3.5          5            0.048544  1.504854
12            4.7          4            0.038835  1.543689
13            3.7          4            0.038835  1.582524
14            4.6          3            0.029126  1.611650
15            3.2          2            0.019417  1.631068
16            3.3          2            0.019417  1.650485
17            3.0          1            0.009709  1.660194
18            3.8          1            0.009709  1.669903

In [70]:
dataset["Overall Rating"].skew()

-0.6570242084406818

In [71]:
dataset["Overall Rating"].kurtosis()

-1.4291371771286214

In [72]:
def Univariate(dataset, quan):
    descriptive=pd.DataFrame(index=["Mean","Median","Mode","Q1:25%","Q2:50%",
                                   "Q3:75%","99%","Q4:100%","IQR","1.5rule","Lesser","Greater","Min","Max","skew","kurtosis"],columns=quan)
    for columnName in quan:
        descriptive[columnName]["Mean"]=dataset[columnName].mean()
        descriptive[columnName]["Median"]=dataset[columnName].median()
        descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
        descriptive[columnName]["Q1:25%"]=dataset.describe()[columnName]["25%"]
        descriptive[columnName]["Q2:50%"]=dataset.describe()[columnName]["50%"]
        descriptive[columnName]["Q3:75%"]=dataset.describe()[columnName]["75%"]
        descriptive[columnName]["99%"]=np.percentile(dataset[columnName],99)
        descriptive[columnName]["Q4:100%"]=dataset.describe()[columnName]["max"]
        descriptive[columnName]["IQR"]=descriptive[columnName]["Q3:75%"]-descriptive[columnName]["Q1:25%"]
        descriptive[columnName]["1.5rule"]=1.5*descriptive[columnName]["IQR"]
        descriptive[columnName]["Lesser"]=descriptive[columnName]["Q1:25%"]-descriptive[columnName]["1.5rule"]
        descriptive[columnName]["Greater"]=descriptive[columnName]["Q3:75%"]+descriptive[columnName]["1.5rule"]
        descriptive[columnName]["Min"]=dataset[columnName].min()
        descriptive[columnName]["Max"]=dataset[columnName].max()
        descriptive[columnName]["skew"]=dataset[columnName].skew()
        descriptive[columnName]["kurtosis"]=dataset[columnName].kurtosis()
    return descriptive  

In [73]:
Univariate(dataset, quan)

Index Overall Rating  Exterior  Interior Ride Quality
Mean        85.5       2.956977  4.468142   4.39823     4.432743
Median      85.5            4.0       4.5       4.4          4.5
Mode           0            0.0       4.5       5.0          4.5
Q1:25%     42.75            0.0       4.2       4.1          4.2
Q2:50%      85.5            4.0       4.5       4.4          4.5
Q3:75%    128.25            4.5       4.9       4.8          4.8
99%       169.29            5.0       NaN       NaN          NaN
Q4:100%    171.0            5.0       5.0       5.0          5.0
IQR         85.5            4.5       0.7       0.7          0.6
1.5rule   128.25           6.75      1.05      1.05          0.9
Lesser     -85.5          -6.75      3.15      3.05          3.3
Greater    256.5          11.25      5.95      5.85          5.7
Min            0            0.0      3.15      3.05          3.3
Max          171            5.0       5.0       5.0          5.0
skew         0.0      -0.657024 -0.905994 -0.738755    -0.759592
kurtosis    -1.2      -1.429137  0.864254   0.62069     0.200236

In [74]:
dataset["Overall Rating"].var()

4.304337005303961

In [75]:
dataset["Overall Rating"].std()

2.0746896166183415

In [76]:
def calculate_descriptive_statistics(dataset, quan):
    descriptive=pd.DataFrame(index=["Mean","Median","Mode","Q1:25%","Q2:50%",
                                   "Q3:75%","99%","Q4:100%","IQR","1.5rule","Lesser","Greater","Min","Max","skew","kurtosis","var","std"],columns=quan)
    for columnName in quan:
        descriptive[columnName]["Mean"]=dataset[columnName].mean()
        descriptive[columnName]["Median"]=dataset[columnName].median()
        descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
        descriptive[columnName]["Q1:25%"]=dataset.describe()[columnName]["25%"]
        descriptive[columnName]["Q2:50%"]=dataset.describe()[columnName]["50%"]
        descriptive[columnName]["Q3:75%"]=dataset.describe()[columnName]["75%"]
        descriptive[columnName]["99%"]=np.percentile(dataset[columnName],99)
        descriptive[columnName]["Q4:100%"]=dataset.describe()[columnName]["max"]
        descriptive[columnName]["IQR"]=descriptive[columnName]["Q3:75%"]-descriptive[columnName]["Q1:25%"]
        descriptive[columnName]["1.5rule"]=1.5*descriptive[columnName]["IQR"]
        descriptive[columnName]["Lesser"]=descriptive[columnName]["Q1:25%"]-descriptive[columnName]["1.5rule"]
        descriptive[columnName]["Greater"]=descriptive[columnName]["Q3:75%"]+descriptive[columnName]["1.5rule"]
        descriptive[columnName]["Min"]=dataset[columnName].min()
        descriptive[columnName]["Max"]=dataset[columnName].max()
        descriptive[columnName]["skew"]=dataset[columnName].skew()
        descriptive[columnName]["kurtosis"]=dataset[columnName].kurtosis()
        descriptive[columnName]["var"]=dataset[columnName].var()
        descriptive[columnName]["std"]=dataset[columnName].std()
    return descriptive

In [77]:
calculate_descriptive_statistics(dataset, quan)

Index Overall Rating  Exterior  Interior Ride Quality
Mean             85.5       2.956977  4.468142   4.39823     4.432743
Median           85.5            4.0       4.5       4.4          4.5
Mode                0            0.0       4.5       5.0          4.5
Q1:25%          42.75            0.0       4.2       4.1          4.2
Q2:50%           85.5            4.0       4.5       4.4          4.5
Q3:75%         128.25            4.5       4.9       4.8          4.8
99%            169.29            5.0       NaN       NaN          NaN
Q4:100%         171.0            5.0       5.0       5.0          5.0
IQR              85.5            4.5       0.7       0.7          0.6
1.5rule        128.25           6.75      1.05      1.05          0.9
Lesser          -85.5          -6.75      3.15      3.05          3.3
Greater         256.5          11.25      5.95      5.85          5.7
Min                 0            0.0      3.15      3.05          3.3
Max               171            5.0       5.0       5.0          5.0
skew              0.0      -0.657024 -0.905994 -0.738755    -0.759592
kurtosis         -1.2      -1.429137  0.864254   0.62069     0.200236
var       2479.666667       4.304337  0.201387  0.222586     0.208472
std         49.796252        2.07469  0.448761   0.47179     0.456587

In [85]:
print(dataset["Overall Rating"].unique())

[4.4 4.2 3.5 3.7 4.8 4.5 5.  4.3 4.7 4.1 3.9 3.2 4.9 4.  3.  4.6 3.8 3.3]


# Since the p-value (0.016) is less than the commonly used significance level of 0.05, we reject the null hypothesis.

# The p-value (in this case, approximately 0.016) is the probability of observing the given result (or a more extreme result) if the null hypothesis is true. In hypothesis testing, the null hypothesis typically assumes that there is no difference between the groups being compared. A low p-value indicates that there is strong evidence against the null hypothesis, suggesting that the observed difference between the groups is statistically significant.

In [87]:
from scipy.stats import ttest_ind

# Filter the dataset for rows where "Overall Rating" equals 4.4 (just as an example)
car_rating = dataset[dataset["Overall Rating"] == 4.4]["Exterior"]
car_rating1 = dataset[dataset["Overall Rating"] == 4.4]["Interior"]
# Perform the t-test
ttest_ind(car_rating, car_rating1)

TtestResult(statistic=2.683281572999736, pvalue=0.016321041449782244, df=16.0)